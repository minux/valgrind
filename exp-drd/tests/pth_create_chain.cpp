// Create threads in such a way that there is a realistic chance that the
// parent thread finishes before the created thread finishes.

#include <cassert>
#include <cstdlib>
#include <iostream>
#include <pthread.h>

static pthread_t s_thread[1000];
static int       s_arg[1000];

static void* thread_func(void* p)
{
  int thread_count = *reinterpret_cast<int*>(p);
  if (thread_count > 0)
  {
    thread_count--;
    // std::cout << "create " << thread_count << std::endl;
    s_arg[thread_count] = thread_count;
    pthread_create(&s_thread[thread_count], 0, thread_func,
		   &s_arg[thread_count]);
#if 0
    std::cout << "created " << thread_count << "(" << s_thread[thread_count]
	      << ")" << std::endl;
#endif
  }
  return 0;
}

int main(int argc, char** argv)
{
  unsigned thread_count = argc > 1 ? atoi(argv[1]) : 50;
  assert(thread_count <= sizeof(s_thread) / sizeof(s_thread[0]));
  assert(thread_count >= 1);
  thread_count--;
  // std::cout << "create " << thread_count << std::endl;
  pthread_create(&s_thread[thread_count], 0, thread_func,
		 const_cast<unsigned*>(&thread_count));
#if 0
  std::cout << "created " << thread_count << "(" << s_thread[thread_count]
	    << ")" << std::endl;
#endif
  for (int i = thread_count; i >= 0; i--)
  {
    // std::cout << "join " << i << "(" << s_thread[i] << ")" << std::endl;
    pthread_join(s_thread[i], 0);
  }
  return 0;
}

// Local variables:
// compile-command: "g++ -o pthread_create-chain -g -Wall -Wextra -Werror -Wno-sign-compare -Wno-unused pthread_create-chain.cpp -lpthread"
// End:
