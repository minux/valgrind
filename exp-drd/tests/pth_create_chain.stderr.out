exp-drd, a data race detector.

Thread 2:
Conflicting load by thread 2 at 0x........ size 8
   at 0x........: (within libpthread-?.?.so)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)
Allocation context: /lib64/libpthread-2.5.so, libpthread.so.0:Text
Other segment start (thread 3)
   (thread finished, call stack no longer available)
Other segment end (thread 3)
   (thread finished, call stack no longer available)

Thread 4:
Conflicting load by thread 4 at 0x........ size 8
   at 0x........: (within /home/sewardj/VgTRUNK/trunk/exp-drd/vgpreload_exp-drd-amd64-linux.so)
   by 0x........: thread_func(void*) (pth_create_chain.cpp:?)
   by 0x........: vg_thread_wrapper (drd_preloaded.c:?)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)
Allocation context: /home/sewardj/VgTRUNK/trunk/exp-drd/vgpreload_exp-drd-amd64-linux.so, NONE:Text
Other segment start (thread 2)
   (thread finished, call stack no longer available)
Other segment end (thread 2)
   (thread finished, call stack no longer available)

Conflicting load by thread 4 at 0x........ size 8
   at 0x........: (within libpthread-?.?.so)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)
Allocation context: /lib64/libpthread-2.5.so, libpthread.so.0:Text
Other segment start (thread 3)
   (thread finished, call stack no longer available)
Other segment end (thread 3)
   (thread finished, call stack no longer available)

Thread 5:
Conflicting load by thread 5 at 0x........ size 8
   at 0x........: (within libpthread-?.?.so)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)
Allocation context: /lib64/libpthread-2.5.so, libpthread.so.0:Text
Other segment start (thread 3)
   (thread finished, call stack no longer available)
Other segment end (thread 3)
   (thread finished, call stack no longer available)

Thread 2:
Conflicting load by thread 2 at 0x........ size 8
   at 0x........: (within libpthread-?.?.so)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)
Allocation context: /lib64/libpthread-2.5.so, libpthread.so.0:Text
Other segment start (thread 3)
   (thread finished, call stack no longer available)
Other segment end (thread 3)
   (thread finished, call stack no longer available)

ERROR SUMMARY: 5 errors from 5 contexts (suppressed: 0 from 0)
